### PR TITLE
4.x closure database

### PR DESCRIPTION
**Pull request by ADmad**
Refs #13611

/cc @lorenzo

### CODE REVIEW

File: src/Database/Query.php [hunk 1]
```diff
@@ -315,7 +316,7 @@ public function sql(?ValueBinder $generator = null): string
      * @param callable $visitor A function or callable to be executed for each part
      * @return $this
      */
-    public function traverse(callable $visitor)
+    public function traverse($visitor)
```
Comment (ADmad) [inline src/Database/Query.php]: Had to remove type hint and rely on parameter widening here as `Query implements ExpressionInterface` and `ExpressionInterface::traverse()` only accepts `Closure` now.

**Comment by ADmad** [comment]
@lorenzo Ping

File: src/Database/Query.php (hunk 1 above)
Comment (markstory) [reply src/Database/Query.php]: Is the concern that userland code is using `traverse()` on a query object?

**Review by markstory — COMMENTED** [review]
Looks good to me as most of the changes effect inward facing APIs.

File: src/Database/Query.php (hunk 1 above)
Comment (ADmad) [reply src/Database/Query.php]: Yes, lorenzo commented [here](https://github.com/cakephp/cakephp/pull/13611#issuecomment-527997495) that the accepted type shouldn't be narrowed down for query methods like formatResults, traverse, map... etc.

**Comment by ADmad** [comment]
phpstan and CS job failures are unrelated. I'll fix them directly on 4.x branch.

**Comment by dereuromark** [comment]
Has this been covered in docs and migration guide yet?

**Comment by ADmad** [comment]
This was mainly about cleaning up the internals of db layer. Most of the function changed are quite low level and don't even have any mention in the docs.

It does affect high level API like `Query::where()` but I don't think anyone would be using array callable or invokable classes are argument for it.

**Pull request merged**